### PR TITLE
UiGraph implementation

### LINKED ISSUE #15: duplicate node index when removing and adding nodes
I haven't figured out what is wrong, but to recreate issue:
1. Start mush
2. Remove any node but the latest created, so `node 0`
3. Add new node, you'll see duplicate node indices on the node labels; in this case two `node 3`'s

### PR DESCRIPTION
**Pull request by stjahns**
Rather than operating directly on the given source graph structure, we construct a new graph structure where each node and edge wraps a corresponding node and edge from the original graph, while also tracking their corresponding widget UiIds

Resolves https://github.com/PistonDevelopers/mush/issues/15

### CODE REVIEW

File: src/toolpane.rs [hunk 1]
```diff
@@ -1,60 +1,32 @@
...
             .dimensions(100.0,40.0)
-            .react(|| {
-                let node = N::default();
-                let position = node.get_position();
-                let node_index = graph.add_node(node);
-                let nuid = (self.nodes.len() + 2) * self.noffset;
-                self.nodes.push(NodeContainer::new(nuid, position, node_index));
-            })
+            .react(|| self.ui_graph.add_node())
             .set(0,ui);
```
Comment (stjahns) [inline src/toolpane.rs:28]: Hm, this should probably use the offset given when ToolPane is initialized

File: src/graph.rs [hunk 2]
```diff
@@ -1,34 +1,157 @@
...
+
+        Button::new()
+            .right(5.0)
+            .label("x")
+            .dimensions(20.0,20.0)
+            .react(|| self.destroy = true)
+            .set(ui_id_start + 2, ui);
+
+        if !self.collapse {
+            Label::new("STUFF")
```
Comment (stjahns) [inline src/graph.rs]: Should probably add `get_label()` to EditableNode and use that

File: examples/demo.rs [hunk 3]
```diff
@@ -20,20 +18,16 @@ use std::path::Path;
...
 
-#[derive(Debug)]
-struct Edge;
-impl EditableEdge for Edge {
-    fn default() -> Self { Edge }
-}
-
-impl EditableNode for NodeState {
+impl EditableNode for DemoNode {
+    fn get_label(&self) -> &str {
```
Comment (viperscape) [inline examples/demo.rs:28]: yes I like this pr, but as this state grows the user would have to impl each new feature, correct? such as get_label. not a huge issue, I think we should merge this either way

Comment (stjahns) [inline examples/demo.rs:28]: We could also add default impls for a lot of the trait methods, if some don't make sense in every situation.

**Pull request merged**